File: 프로그래머스/2/152996. 시소 짝꿍/시소 짝꿍.java
import java.util.*;
class Solution {
    public long solution(int[] weights) {
        long answer = 0;
        
        Map<Double, Integer> weightCount = new HashMap<>();
        Set<Double> weightSet = new HashSet();
        
        for (int w : weights) {
            weightCount.put(w * 1.0, weightCount.getOrDefault(w * 1.0, 0) + 1);
            weightSet.add(w * 1.0);
        }
        
        for (Double w : weightSet) {
            if (weightCount.get(w) >= 2) {
                answer +=  ((long) weightCount.get(w) * (long) (weightCount.get(w) - 1)) / 2;
            }
            
            if (weightCount.containsKey(w * 3 / 2)) {
                answer += (long) weightCount.get(w*3/2) * (long) weightCount.get(w);
            }
            if (weightCount.containsKey(w * 4 / 2)) {
                answer += (long) weightCount.get(w*4/2) * (long) weightCount.get(w);
            }
            if (weightCount.containsKey(w * 4 / 3)) {
                answer += (long) weightCount.get(w*4/3) * (long) weightCount.get(w);
            }
        }
 
        return answer;
    }
}

File: 프로그래머스/2/152996. 시소 짝꿍/시소 짝꿍.py
def solution(weights):
    answer = 0
    
    info = dict()
    for w in weights:
        if w in info:
            info[w] += 1
        else:
            info[w] = 1
            
    for k, v in info.items():
        if v >= 2:
            answer += (v * (v-1)) // 2
    
    weights = set(weights)
    # 2:3, 2:4, 3:4 비율을 가지면 짝꿍 가능
    for w in weights:
        if w * 2 / 3 in weights:
            answer += info[w*2/3] * info[w]
        if w * 2 / 4 in weights:
            answer += info[w*2/4] * info[w]
        if w * 3 / 4 in weights:
            answer += info[w*3/4] * info[w]
    return answer